src/Api/Controllers/AuthRequestsController.cs

```diff
@@ -0,0 +1,155 @@
+        var authRequest = await _authRequestRepository.GetByIdAsync(new Guid(id));
+        if (authRequest == null || authRequest.UserId != userId)
+        {
+            throw new NotFoundException();
+        }
+
+        return new AuthRequestResponseModel(authRequest, _globalSettings.SelfHosted);
+    }
+
+    [HttpGet("{id}/response")]
+    [AllowAnonymous]
+    public async Task<AuthRequestResponseModel> GetResponse(string id, [FromQuery] string code)
+    {
+        var authRequest = await _authRequestRepository.GetByIdAsync(new Guid(id));
+        if (authRequest == null || code != authRequest.AccessCode)
+        {
+            throw new NotFoundException();
+        }
+
+        if (authRequest.GetExpirationDate() < DateTime.UtcNow)
+        {
+            throw new NotFoundException();
+        }
+
+        return new AuthRequestResponseModel(authRequest, _globalSettings.SelfHosted);
+    }
+
+    [HttpPost("")]
+    [AllowAnonymous]
+    public async Task<AuthRequestResponseModel> Post([FromBody] AuthRequestCreateRequestModel model)
+    {
+        var user = await _userRepository.GetByEmailAsync(model.Email);
+        if (user == null)
+        {
+            throw new NotFoundException();
+        }
+        if (!_currentContext.DeviceType.HasValue)
+        {
+            throw new BadRequestException("Device type not provided.");
+        }
+        if (!_globalSettings.PasswordlessAuth.KnownDevicesOnly)
+        {
+            var d = await _deviceRepository.GetByIdentifierAsync(_currentContext.DeviceIdentifier);
+            if (d == null || d.UserId != user.Id)
+            {
+                throw new NotFoundException();
+            }
+        }
+
+        var authRequest = new AuthRequest
+        {
+            RequestDeviceIdentifier = model.DeviceIdentifier,
+            RequestDeviceType = _currentContext.DeviceType.Value,
+            RequestIpAddress = _currentContext.IpAddress,
+            AccessCode = model.AccessCode,
+            PublicKey = model.PublicKey,
+            UserId = user.Id,
+            Type = model.Type.Value,
+            RequestFingerprint = model.FingerprintPhrase
+        };
+        authRequest = await _authRequestRepository.CreateAsync(authRequest);
+        await _pushNotificationService.PushAuthRequestAsync(authRequest);
+        return new AuthRequestResponseModel(authRequest, _globalSettings.SelfHosted);
+    }
+
+    [HttpPut("{id}")]
+    public async Task<AuthRequestResponseModel> Put(string id, [FromBody] AuthRequestUpdateRequestModel model)
+    {
+        var userId = _userService.GetProperUserId(User).Value;
+        var authRequest = await _authRequestRepository.GetByIdAsync(new Guid(id));
+        if (authRequest == null || authRequest.UserId != userId)
+        {
+            throw new NotFoundException();
+        }
+
+        if (authRequest.GetExpirationDate() < DateTime.UtcNow)
+        {
+            throw new NotFoundException();
+        }
```

Comment:
Same question as above in the Get about combining these

Reply:
Implemented on [[review] Consolidate error conditions](https://github.com/bitwarden/server/pull/2276/commits/2853331550a040068c3af68ae6067ee559f63281)